File: 序号/1552. 两球之间的磁力.py
class Solution:
    def maxDistance(self, position: List[int], target: int) -> int:
        # 最小可能间距1，最大可能间距position[-1]-posision[0]
        # 二分找ans，从左到右扫position，看间距ans能否满足放得下m
        position.sort()
        n,l,r=len(position),0,position[-1]-position[0]+1
        
        while l+1<r:
            m = (l+r)//2
            cnt,pre=1,position[0]
            for i in range(1,n):
                if position[i]-pre>=m:
                    cnt+=1
                    pre = position[i]
            if cnt>=target:l=m
            else: r=m
        return l

File: 序号/1552. 两球之间的磁力.kt
class Solution {
  fun maxDistance(position: IntArray, m: Int): Int {
      // 最小最大问题，无脑二分
      position.sort()
      val limit = position.last()-position.first()+1
      var l = 0
      var r = limit
      while(l+1<r){
          val mid = (l+r) shr 1 // 判断两球间隔m 可不可以
          var cnt = 1
          var prev = position[0]
          for(i in 1 until position.size){
              val diff = position[i] - prev
              if(diff >= mid){
                  if(++cnt==m) break // ok，所有球都放下了
                  prev = position[i] // 满足条件，找下一个球
              }
          }
          if(cnt==m){
              l=mid
          }else{
              r=mid
          }

      }
      return l
  }
}